src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -350,12 +351,46 @@ private DocumentState SetParseOptions(ParseOptions options)
                 throw new InvalidOperationException();
             }
 
-            var newTreeSource = CreateLazyFullyParsedTree(
-                TextAndVersionSource,
-                Id.ProjectId,
-                GetSyntaxTreeFilePath(Attributes),
-                options,
-                _languageServices);
+            ValueSource<TreeAndVersion> newTreeSource;
+
+            // Optimization: if we are only changing preprocessor directives, and we've already parsed the existing tree and it didn't have
+            // any, we can avoid a reparse since the tree will be parsed the same.
+            if (onlyPreprocessorDirectiveChange &&
+                _treeSource.TryGetValue(out var existingTreeAndVersion) &&
+                existingTreeAndVersion.Tree.TryGetRoot(out var existingRoot) &&
+                !existingRoot.ContainsDirectives)
```

Comment:
does this only work if the trees are in memory/  if not in memory we have to dump and recreate it?   any way we might be able to cache this and store it around any time we create a new root?

Reply:
Possibly, see point 3 in the PR text. Given the project system needs to do additional work here before this really lights up, so I didn't want to add too much complexity here yet since this is still somewhat speculative. I guess I should add a clear TODO here for that case, at least.

Reply:
ah yes.  i missed that.  and this does seem really unfortunate.  if you test this with rolsyn, how well does it work?  i feel like we'll likely have dumped large swaths of hte solution, so we still have to pay for a ton of this work.  Is that not hte case?

Reply:
So I can't test with *current* Roslyn because we need CPS to fix their side first. I can try to find some old version of Roslyn that might work.

If you want me to front-load that I can do it; I also couldn't remember if you were currently A/B testing that and might get rid of it soon...?

Reply:
@CyrusNajmabadi OK, the recoverable side is also implemented now too, and at least tested in unit tests.